pkg/ottl/functions.go

```diff
@@ -93,116 +98,126 @@ func (p *Parser[K]) buildArgs(inv invocation, fType reflect.Type) ([]reflect.Val
+			return nil, fmt.Errorf("invalid argument at position %v %w", index, err)
 		}
-		*args = append(*args, reflect.ValueOf(arg))
+		return arg, nil
 	case strings.HasPrefix(name, "Getter"):
 		arg, err := p.newGetter(argDef)
 		if err != nil {
-			return fmt.Errorf("invalid argument at position %v %w", index, err)
+			return nil, fmt.Errorf("invalid argument at position %v %w", index, err)
 		}
-		*args = append(*args, reflect.ValueOf(arg))
+		return arg, nil
 	case name == "Enum":
 		arg, err := p.enumParser(argDef.Enum)
 		if err != nil {
-			return fmt.Errorf("invalid argument at position %v must be an Enum", index)
+			return nil, fmt.Errorf("invalid argument at position %v must be an Enum", index)
 		}
-		*args = append(*args, reflect.ValueOf(*arg))
+		return *arg, nil
 	case name == "string":
 		if argDef.String == nil {
-			return fmt.Errorf("invalid argument at position %v, must be an string", index)
+			return nil, fmt.Errorf("invalid argument at position %v, must be an string", index)
 		}
-		*args = append(*args, reflect.ValueOf(*argDef.String))
+		return *argDef.String, nil
 	case name == "float64":
 		if argDef.Float == nil {
-			return fmt.Errorf("invalid argument at position %v, must be an float", index)
+			return nil, fmt.Errorf("invalid argument at position %v, must be an float", index)
 		}
-		*args = append(*args, reflect.ValueOf(*argDef.Float))
+		return *argDef.Float, nil
 	case name == "int64":
 		if argDef.Int == nil {
-			return fmt.Errorf("invalid argument at position %v, must be an int", index)
+			return nil, fmt.Errorf("invalid argument at position %v, must be an int", index)
 		}
-		*args = append(*args, reflect.ValueOf(*argDef.Int))
+		return *argDef.Int, nil
 	case name == "bool":
 		if argDef.Bool == nil {
-			return fmt.Errorf("invalid argument at position %v, must be a bool", index)
+			return nil, fmt.Errorf("invalid argument at position %v, must be a bool", index)
 		}
-		*args = append(*args, reflect.ValueOf(bool(*argDef.Bool)))
+		return bool(*argDef.Bool), nil
 	default:
-		return errors.New("unsupported argument type")
+		return nil, errors.New("unsupported argument type")
 	}
-	return nil
 }
 
 // Handle interfaces that can be declared as parameters to a OTTL function, but will
 // never be called in an invocation. Returns whether the arg is an internal arg.
-func (p *Parser[K]) buildInternalArg(argType reflect.Type, args *[]reflect.Value) bool {
+func (p *Parser[K]) buildInternalArg(argType reflect.Type) (reflect.Value, bool) {
 	switch argType.Name() {
 	case "TelemetrySettings":
-		*args = append(*args, reflect.ValueOf(p.telemetrySettings))
+		return reflect.ValueOf(p.telemetrySettings), true
 	default:
-		return false
+		return reflect.ValueOf(nil), false
+	}
+}
+
+type BuildArg func(value, reflect.Type, int) (any, error)
+
+func buildSlice[T any](inv invocation, argType reflect.Type, index int, buildArg BuildArg, name string) (reflect.Value, error) {
```

Comment:
Can build slice be a member function of Parser so that we can directly reference `buildArg` instead of having to pass it as a function?

Reply:
Unfortunately methods can't be generic, which is why this this is just a function.

Reply:
How messy do things get if we make it a member function instead of using generics?  Do we end up with a method per type we support?

Reply:
Yeah, we would end up with a method per supported slice type. I played around with this a bit and couldn't find a solution I liked more.

Reply:
> up with a method per supported slice type

Ya I don't love this either.  Can the new type be unexported?

@bogdandrutu @kentquirk do you have any opinions?

Reply:
Thanks for catching that, didn't mean to export it. It's package-scoped now.